experimental/ecma402_listformat/src/lib.rs

```diff
@@ -0,0 +1,554 @@
+    ///
+    /// See [Options] for the contents of the options.  See the [Format::try_new]
+    /// for the use of the options.
+    pub mod options {
+        /// Chooses the list formatting approach.
+        #[derive(Eq, PartialEq, Debug, Clone)]
+        pub enum Style {
+            Long,
+            Short,
+            Narrow,
+        }
+        /// Chooses between "this, that and other", and "this, that or other".
+        #[derive(Eq, PartialEq, Debug, Clone)]
+        pub enum Type {
+            /// "This, that and other".
+            Conjunction,
+            /// "This, that or other".
+            Disjunction,
+        }
+    }
+
+    /// The options set by the user at construction time.  See discussion at the top level
+    /// about the name choice.  Provides as a "bag of options" since we don't expect any
+    /// implementations to be attached to this struct.
+    ///
+    /// The default values of all the options are prescribed in by the [TC39 report][tc39lf].
+    ///
+    ///   [tc39lf]: https://tc39.es/proposal-intl-list-format/#sec-Intl.ListFormat
+    #[derive(Eq, PartialEq, Debug, Clone)]
+    pub struct Options {
+        /// Selects a [options::Style] for the formatted list.  If unset, defaults
+        /// to [options::Style::Long].
+        pub style: options::Style,
+        /// Selects a [options::Type] for the formatted list.  If unset, defaults to
+        /// [options::Type::Conjunction].
+        pub in_type: options::Type,
+    }
+
+    /// Allows the use of `listformat::Factory::try_new(..., Default::default())`.
+    impl Default for Options {
+        /// Gets the default values of [Options] if omitted at setup.  The
+        /// default values are prescribed in by the [TC39 report][tc39lf].
+        ///
+        ///   [tc39lf]: https://tc39.es/proposal-intl-list-format/#sec-Intl.ListFormat
+        fn default() -> Self {
+            Options { style: options::Style::Long, in_type: options::Type::Conjunction }
+        }
+    }
+
+    use std::fmt;
+
+    /// The package workhorse: formats supplied pieces of text into an ergonomically formatted
+    /// list.
+    ///
+    /// While ECMA 402 originally has functions under `Intl`, we probably want to
+    /// obtain a separate factory from each implementor.
+    ///
+    /// Purposely omitted:
+    ///
+    /// - `supported_locales_of`.
+    pub trait Format<'a> {
+        /// A better name for this associated type is most welcome.
+        type Ret: Format<'a>;
+
+        /// The type of error reported, if any.
+        type Error: std::error::Error;
+
+        /// Creates a new [Format].
+        ///
+        /// Creation may fail, for example, if the locale-specific data is not loaded, or if
+        /// the supplied options are inconsistent.
+        fn try_new(l: impl crate::Locale, opts: Options) -> Result<Self::Ret, Self::Error>;
+
+        /// The type used to implement [crate::Locale].
+        type Locale: crate::Locale + 'a;
```

Comment:
@Manishearth - can you take another peak at the `type Locale`, type LocaleIter` and `type Ret`? Does it look like the right API surface?